### DISCUSSION TITLE
Clean Architecture no frontend

### DISCUSSION THREAD

**Discussion by iamdivaloper** [frontendbr/forum#1600] [Discussão]
Já ouviram falar sobre? Já tiveram que implementar o padrão de arquitetura limpa introduzido pelo Uncle Bob para aplicativos da Web como o Angular? Se sim, pode contar como foi sua experiencia, os prós e contras. 

Referências:
[Clean Architecture - Uncle Bob](http://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html)
[Clean Architecture - Dataflow](https://proandroiddev.com/clean-architecture-data-flow-dependency-rule-615ffdd79e29)
[Nomenclaturas e boas práticas](https://blog.codeminer42.com/nodejs-and-good-practices-354e7d763626)
[Use Cases](http://www.plainionist.net/Implementing-Clean-Architecture-UseCases/)
[Clean Frontend Architecture](https://dev.to/phodal/clean-architecture-for-frontend-in-action-1aop)

https://github.com/im-a-giraffe/angular-clean-architecture
https://github.com/leandro-mancini/angular-clean-architecture

**Comment by fabiomirandaa** [comment]
Sinceramente, eu não uso ainda mas eu gosto bastante do conceito. Inclusive, dei uma olhada aqui no Repositório da arquitetura aplicada no angular que você colocou e achei sensacional. 

Atualmente eu uso algo mais ou menos assim:
```
app/
|- core/
   |- auth/
   |- home/
      |- pages/
      |- services/
      |- components/
      |- interfaces/
|- shared/
   |- shared.module.ts
   |- components/
      |- bagulhete-reutilizavel/
         |- bagulhete-reutilizavel.component.ts
         |- bagulhete-reutilizavel.component.css
   |- pipes/
(continuaria abaixo com services e etc)
```

**Comment by fabiomirandaa** [comment]
Nossa, mais ninguém? Queria muito ver essa discussão aqui rolando...

**Comment by evelisee-zz** [comment]
Eu já utilizei em uma aplicação Design System e em outros projetos com grandes regras de negócio  (validação de formulário gigantescas). Sinceramente, foi uma das melhores opções pra desenvolver de forma inteligente onde cada bloco de código deve ficar e principalmente dividindo responsabilidades. Segue um print da estrutura inicial.

O meu feedback é que dá bastante trabalho (mais que outras arquiteturas) porque força o desenvolvedor a seguir os princípios de SOLID, mas por outro lado facilita quando obtemos ajuda de pessoas de outras áreas que querem mexer no código. Uma vez que essa arquitetura está bem persistente no back-end, torna os projetos, mesmo separados, conversando em uma linguágem única.

Aqui tem uma representação bem bacana do gráfico que mostra a divisão das camadas e um print da arquitetura aplicada em um projeto angular:
![Screenshot from 2019-11-21 14-14-17](https://user-images.githubusercontent.com/10232707/69360404-3e078100-0c69-11ea-8bb2-acb24c9c7782.png)

**Reply by alysonfs** [reply]
Ola @evelisee-zz , Eu estou tentando aplica o clean code. Vou fizer, é muito bom. Entretanto, quando estou para implementar uma nova funcionalidade, sempre tenho vários questionamentos, é muito difícil, para quem tá começando, saber onde as coisas de encaixam. O livro está aqui na minha frente, mas as demandas atropelam. 
Gostaria de te perguntar sobre, em uma projeto full, você aplica mas mesmas regras de negócio no front e no Back? Existe uma forma de reduzir esse custo de programação, pensado em criar um meio "Core" da aplicação onde o front e o Back compartilhem as identidades de casos de uso? OU não, no front tenho que refazer as camadas idênticas ao Back. 
Esse forum está aberto a bastante tempo, 2 anos se passaram, talvez vc já tenha encarado uma batalha dessa.
Obrigado por compartilhar.

**Comment by kivervinicius** [comment]
> Eu já utilizei em uma aplicação Design System e em outros projetos com grandes regras de negócio (validação de formulário gigantescas). Sinceramente, foi uma das melhores opções pra desenvolver de forma inteligente onde cada bloco de código deve ficar e principalmente dividindo responsabilidades. Segue um print da estrutura inicial.
> 
> O meu feedback é que dá bastante trabalho (mais que outras arquiteturas) porque força o desenvolvedor a seguir os princípios de SOLID, mas por outro lado facilita quando obtemos ajuda de pessoas de outras áreas que querem mexer no código. Uma vez que essa arquitetura está bem persistente no back-end, torna os projetos, mesmo separados, conversando em uma linguágem única.
> 
> Aqui tem uma representação bem bacana do gráfico que mostra a divisão das camadas e um print da arquitetura aplicada em um projeto angular:
> ![Screenshot from 2019-11-21 14-14-17](https://user-images.githubusercontent.com/10232707/69360404-3e078100-0c69-11ea-8bb2-acb24c9c7782.png)

um pouco fora do contexto, mas pensei que somente nossa empresa tivesse problema de misturar entidades PT-BR com EN

**Comment by evelisee-zz** [comment]
> > Eu já utilizei em uma aplicação Design System e em outros projetos com grandes regras de negócio (validação de formulário gigantescas). Sinceramente, foi uma das melhores opções pra desenvolver de forma inteligente onde cada bloco de código deve ficar e principalmente dividindo responsabilidades. Segue um print da estrutura inicial.
> > O meu feedback é que dá bastante trabalho (mais que outras arquiteturas) porque força o desenvolvedor a seguir os princípios de SOLID, mas por outro lado facilita quando obtemos ajuda de pessoas de outras áreas que querem mexer no código. Uma vez que essa arquitetura está bem persistente no back-end, torna os projetos, mesmo separados, conversando em uma linguágem única.
> > Aqui tem uma representação bem bacana do gráfico que mostra a divisão das camadas e um print da arquitetura aplicada em um projeto angular:
> > ![Screenshot from 2019-11-21 14-14-17](https://user-images.githubusercontent.com/10232707/69360404-3e078100-0c69-11ea-8bb2-acb24c9c7782.png)
> 
> um pouco fora do contexto, mas pensei que somente nossa empresa tivesse problema de misturar entidades PT-BR com EN

HAHAHAHAHA sempre tem, está no nosso dia a dia!
